amin.1
.\"	@(#)amin.1	2.41 (Wolter) 1/27/2000
.\"
.TH AMIN 1 "July 1, 1991"
.AT 3
.SH NAME
amin \- notify writers that you are busy
.SH SYNOPSIS
.B amin
[-ynesp] command [args...]
.SH DESCRIPTION
.I Amin
is used when you don't want to be written while running a command.
It runs the command given normally.  If your message permissions
(see
.IR mesg (1))
are off, it does nothing much else.
If your messages are on, people writing you with
.IR write (1)
will be warned that you are running that command and
will be given the opportunity to change their minds about writing you.
.PP
The
.B -n
option may be used to turn your messages entirely off for the duration
of the execution of command.
People writing you will get "Permission denied".
The
.B -y
option turns your message permissions on for the duration of
the execution of the command.
The
.B -e
may be used after either
.B -n
or
.B -y
to indicate that the logins listed in the
.I .yeswrite
or the
.I .nowrite
files respectively are exceptions to the message permissions set.
The default is
.B -s
which leaves your message permissions in their original state.
In any case, after the command is complete,
your permissions will be restored to the original state.
.PP
The
.B -p
flag causes all telegrams sent to you while the command is running to be saved.
They are displayed as soon as the command is complete.
If used with the
.B -n
flag, writes are refused, but telegrams are still saved.
.PP
If you have designated yourself as a helper, you will still be marked on
the
.IR finger (1)
output as a helper while you are running
.I amin
but people doing ``write help'' will not be connected to you,
even if you have the helper flag set to ``Y''.
.SH AUTHOR
Jan Wolter
.SH FILES
/etc/wrttmp	to find message permissions
.br
/etc/utmp	to find user
.SH "SEE ALSO"
mesg(1), finger(1), write(1), huh(1).


CHANGES
 Version 2.55:   Sep 28, 2004
   - Rename log() function to wrtlog() to avoid conflicts with math.h log()
     function.
   - Add distclean target to Makefile.

 Version 2.54:   May 30, 2003
   - Fixed a couple buffer overrun problems that might have constituted a
     possible security hole.
   - Small portability fixes.

 Version 2.53:   Jul 6, 2000
   - Minor portability improvements to install scripts.

 Version 2.52:   Jun 19, 2000
   - Fixed a buffer overflow possibility in huh.c.  This is a major problem if
     'huh' is installed as root.  Older versions should not be installed that
     way.

 Version 2.51:   Feb 20, 2000
   - Renamed lib_utmp.c as getutent.c, and cleaned it up to work more similarly
     to the standard getutent() functions.  Fixed some bugs Karyl Stein pointed
     out with Linux utmp files.

 Version 2.50: Jan 28, 2000
   - Add -pk and -x* flags to mesg, so if we have a cooperative talkd, we can
     control talk permissions seperately from write/tel permissions.
   - Can optionally include hostnames in "Message from" headers.
   - If a user is logged on more than once, write least idle tty.
   - Use autoconf to generate ./configure script which makes Makefile and
     and config.h.
   - Replace old wrttab file with new orville.conf file, and move most site
     configuration into there.
   - Include timezone in message header, since sender's timezone is not
     necessarily the same as the recipients.
   - Lots of rearrangements of header files and such.
   - Miscellaneous protability improvements.
   - Write -r option supported for more different kinds of password files.
   - Changed default name of the 'write -c' link to write from 'chat' to 'jot'.
     Chat is a better name, and I had it first, but the other chat is more
     widely distributed.

 Version 2.41:  Mar 17, 1999
   - ANSIfication of all function declarations.
   - Use getutent() on systems that have it.  Need this when a utmp daemon
     is used instead of a utmp file (thanks to Shane Wegner <shane@cm.nu>).
   - Be a bit more careful about printing/deleting .lastmesg to ensure we
     aren't root when we do it.
   - clean target added to Makefile (thanks to Shane Wegner <shane@cm.nu>).
   - Renamed "common.c" to "lib_common.c" because I wanted to.
   - Lots of minor code cleanup.

 Version 2.40:
   - "mesg -bn" turns off bells.
   - Fixed bug in disconnect so it disconnects multiple writes.
   - exceptions through .yeswrite and .nowrite files and "mesg ye/ne" setting
   - telegram mode no longer sets wrt_what, since it is too transient.
   - telegram mode ignores -p, -y, -n, -c, -l flags (unless switch to write)
   - added helpers command.
   - Day of week in log file dates.
   - mesg command prints warning if people still can write you.
   - If previous helper is still available, another "write help" selects him.
   - Disable core dumps and check that stderr is open.
   - Security patches to .lastmesg file creation. [e]
   - Security patches to .lastmesg reading. [f]
   - Portability improvements to input flushing code

 Version 2.34:
   - Sizes of wrttmp entries stored in wrttmp header.
   - Can write back to someone writing you even if he prefers tels.
   - "mesg d" is like "mesg N" but doesn't change permissions.
   - Linux bug fixes.
   - Made writhist more resistant to system clock errors.

 Version 2.33:
   - Include uid name in message banner if different from utmp name.
   - Root can override recipient's write/tel preferences.
   - Optional recording of telegrams (with mesg -r, amin -p and huh)
   - "mesg -hY" makes you a helper even when permissions are off.

 Version 2.32:
   - wrthist file added, permitting tel replys and regulating flooding.
   - expand control-characters on local echo as well as send.
   - teach it to echo tabs and backspaces more accurately.
   - some low-grade word-wrap in character mode.
   - HELPLOG option for logging help requests only.

 Version 2.31:
   - "write ." writes again to the last user written.
   - write/telegram and line/character mode preferences added.
   - support for login names with more than 8 characters.
   - optional logging of all write requests.
   - added -v to mesg. [2.31e]
   - NOHELPER defines a file to print if there are no helpers. [2.31f]

 Version 2.30:
   - Integration of telegram option into write.
   - Fix bug with shell escapes in "write -n" making "mesg n" perminant.
   - Shell escapes use user's SHELL instead of always using /bin/sh.

 Version 2.20:
   - Reorganization of source code for greater modularity.
   - If multiple helpers are available, choose one at random.
   - When input is redirected, don't ask use stdin to ask user if he wants
     to interupt a conversation.
   - Don't go into character mode till other person has replied.
   - Starting a line with ) in character mode does that line in line mode.
   - If compiled with the NOVICEHELP variable and run with the NOVICE
     environment variable set, prints the contents of the file named in
     NOVICEHELP before running.
   - Fixed amin's job control.

 Version 2.11:
   - If user is logged on more than once, prefer tty that is permitted.
   - Clearer "Permission denied" message.

 Version 2.12:
   - Added TTY_OTHERS

 Version 2.10:
   - Attempted SYSV Compatible system.
   - Added HELPERS

 Version 2.00:  First BSD Compatible system.
   - Added wrttab.
   - Added consistancy checking between utmp and wrttmp to eliminate need for
     modifications to /bin/login.
   - Added support for job control and other BSD-isms.

 Version 1.05:  Last M-Net system III specific version.

 Version 1.00:  First release - Sept 1985.


amin.c
/* AMIN.C
 *
 * Run a command, warning potential writers that you are in it.
 *
 *  8-24-86  Written by Jan Wolter
 *  8-27-86  Added code to make it work with shell scripts
 *  2-12-94  Fixed suspends.
 * 11-30-95  Added -p
 */

#include "orville.h"
#include <signal.h>
#include <errno.h>
#include <pwd.h>
#ifdef HAVE_SYS_WAIT_H
# include <sys/wait.h>
#endif
#ifndef WEXITSTATUS
#define WEXITSTATUS(x) ((unsigned)(x) >> 8)
#endif
#ifndef WIFEXITED
#define WIFEXITED(x) (((x) & 255) == 0)
#endif

#include <sys/stat.h>

extern int errno;

char *my_dir(void);
RETSIGTYPE susp(void);
void fix_modes(char *fullcmd);
void restore_modes(void);
int exec_cmd(int argc, char **argv);

struct wrttmp mywrt;	/* My old wrttmp entry, to be restored at exit time */
struct wrttmp newwrt;	/* Space to build temporary wrttmp entry */
long mypos= -1;		/* offset of my entry in the wrttmp file */
char tmp_mesg= 's';	/* Temporary setting of mesg during write */
int fixed_modes= 0;	/* Have we fixed the wrttmp/tty modes? */
int postpone= 0;	/* Should telegrams be postponed? */
int exceptions= 0;	/* Should we allow exceptions on message permissions */


int main(int argc, char **argv)
{
char cmdbuf[280];	/* Space to build full command path name */
char lmfile[200];	/* Space to store full path name of .lastmesg file */
FILE *lmfp= NULL;
int i,rc;
char *c;

    progname= leafname(argv[0]);
    readconfig(NULL);

    for (i= 1; i<argc; i++)
    {
	if (*(c= argv[i]) == '-')
	    while (*(++c) != '\000')
		switch(*c)
		{
		case 'n':
		case 'y':
		case 's':
			tmp_mesg= *c;
			break;
		case 'p':
			postpone= 1;
			break;
		case 'e':
			if (!f_exceptions)
			{
			    printf("%s: exceptions are not enabled on this "
				   "system so -e cannot be used.\n",progname);
			    exit(1);
			}
			if (tmp_mesg != 'n' && tmp_mesg != 'y')
			{
			    printf("%s: -e option must be preceded by "
				   "-n or -y option.\n",progname);
			    exit(1);
			}
			exceptions= 1;
			break;
		default:
			goto usage;
		}
	else 
	    break;
    }
    argv= argv+i;
    argc= argc-i;

    if (argc < 1) goto usage;

    /* Trap darn near every possible signal */
    signal(SIGINT,SIG_IGN);
    signal(SIGQUIT,SIG_IGN);
    signal(SIGTERM,SIG_IGN);
    signal(SIGHUP,SIG_IGN);
#ifdef SIGTSTP
    signal(SIGTSTP,(RETSIGTYPE (*)())susp);
#endif /*SIGTSTP*/

    /* Open the wrttmp file */
    if (init_wstream(O_RDWR))
	exit(1);

    /* Give up my root status */
    setuid(getuid());
    setgid(getgid());

    /* Check if device 2 is a tty open for read */
    if (!(fcntl(2,F_GETFL,0) & O_RDWR) || !isatty(2))
    {
	printf("%s: I/O stream 2 improperly redirected\n",progname);
	exit(1);
    }

    if (postpone)
    {
	char *dir;
	struct stat st;

	if ((dir= my_dir()) == NULL)
	    postpone= 0;
	else
	{
	    sprintf(lmfile,"%.180s/.lastmesg",dir);
	    unlink(lmfile);
	}
    }

    fix_modes(argv[0]);

    /* Fork off child */
    rc= exec_cmd(argc,argv);

    if (postpone && (lmfp= fopen(lmfile, "r")) != NULL)
	unlink(lmfile);

    restore_modes();

    signal(SIGINT,SIG_DFL);
    signal(SIGQUIT,SIG_DFL);
    signal(SIGTERM,SIG_DFL);
    signal(SIGHUP,SIG_DFL);
#ifdef SIGTSTP
    signal(SIGTSTP,SIG_DFL);
#endif

    if (lmfp != NULL)
    {
	int ch;

	printf("Messages received while you were busy...\n");
	while ((ch= getc(lmfp)) != EOF)
	    if (ch != '\007') putchar(ch);
    }

    exit(rc);

usage:
    printf("usage: %s [-ny%ssp] <command> <args>\n",progname,
    	f_exceptions ? "e" : "");
    exit(1);
}


/* MY_DIR - Return name of my home directory, or NULL if I can't find it.
 */

char *my_dir()
{
struct passwd *pwd;
char *r;

    /* Try the environment variable first */
    if ((r= getenv("HOME")) != NULL) return(r);

    /* Try the password file */
    setpwent();
    pwd= getpwuid(getuid());
    endpwent();
    if (pwd != NULL)
	return(pwd->pw_dir);

    /* Give up */
    return(NULL);
}


/* LOCATE_WRTTMP - find my entry in the wrttmp file.
 */

void locate_wrttmp(char *tty, struct wrttmp *wbuf, long *pos)
{
struct utmpx *ut;

    /* Find utmp entry */
    if ((ut= find_utmp(tty)) == NULL || ut->ut_user[0] == '\0')
    {
	printf("%s: Can't find your tty (%s) in utmp\n",progname,tty);
	exit(1);
    }

    find_wrttmp(tty, ut->ut_tv.tv_sec, wbuf, pos);
}


/* FIX_MODES - modify wrttmp and tty permissions.
 */

void fix_modes(char *fullcmd)
{
char *mytty;		/* my tty name in "tty##" format */
char *shortcmd;		/* command without full pathname */

    /* Find my tty line */
    if (getdevtty()) exit(1);
    mytty= mydevname+5;

    /* Find my wrttmp entry */
    locate_wrttmp(mytty,&mywrt,&mypos);

    if (mypos < 0)
    {
	printf("%s: Error - unable to locate %s in %s\n", 
		progname, mytty, f_wrttmp);
	exit(1);
    }

    /* Close utmp file */
    endutxent();

    /* Figure out name of program being exec'ed */
    if ((shortcmd= strrchr(fullcmd,'/')) == NULL)
	shortcmd= fullcmd;
    else
	shortcmd++;

    /* Fix my entry in wrttmp */
    newwrt= mywrt;
    newwrt.wrt_what[0]= '!';
    strncpy(newwrt.wrt_what+1, shortcmd, sizeof(((struct utmpx *)0)->ut_user)-2);
#ifndef TTYPERMS
    if (tmp_mesg != 's') newwrt.wrt_mesg= tmp_mesg;
#endif
    if (tmp_mesg != 's')
    	newwrt.wrt_except= (f_exceptions && exceptions) ? 'y' : 'n';
    if (postpone) newwrt.wrt_record= 'a';
    lseek(wstream,mypos,0);
    write(wstream,&newwrt,sizeof(struct wrttmp));

#ifdef TTYPERMS
    /* Fix my tty permissions */
    if (tmp_mesg != 's')
    {
	    if (saveperms()) exit(1);
	    setperms(tmp_mesg);
    }
#endif /*TTYPERMS*/
    fixed_modes= 1;
}


/* RESTORE_MODES - Restore original modes.
 */

void restore_modes()
{
    /* Restore old (or default) wrttmp entry */
    lseek(wstream,mypos,0);
    write(wstream,&mywrt,sizeof(struct wrttmp));
    close(wstream);

#ifdef TTYPERMS
    if (tmp_mesg != 's') resetperms();
#endif /*TTYPERMS*/
}


/* EXEC_CMD - Run the command as a child process
 */

int exec_cmd(int argc, char **argv)
{
int cpid,wpid;		/* Child pid and pid returned by wait */
char **shargv;		/* Arguements for running things with shell */
int status;		/* Child's return status */
int i;

    if ((cpid= fork()) == 0)
    {
	/* Child process */

	/* Use default interupts */
	signal(SIGINT,SIG_DFL);
	signal(SIGQUIT,SIG_DFL);
	signal(SIGTERM,SIG_DFL);
	signal(SIGHUP,SIG_DFL);
#ifdef SIGTSTP
	signal(SIGTSTP,SIG_DFL);
#endif /*SIGTSTP*/

	/* Close open file descriptors */
	/* close(wstream);*/  /* Don't need this, since set close-on-exec */
	endutent();	/* Might or might not need this */

	execvp(argv[0],argv);
	/* If wasn't executable, try running under a shell */
	if (errno == ENOEXEC)
	{
		shargv= (char **)malloc((argc+3)*sizeof(char *));
		shargv[0]= "sh";
		shargv[1]= "-c";
		for (i= 0; i < argc; i++)
		    shargv[i+2]= argv[i];
		shargv[i+2]= NULL;
		execv("/bin/sh",shargv);
	}
	
	printf("%s: command not found\n",argv[0]);
	exit(1);
    }
    /* Parent Process waits for child to exit */
    while ((wpid= wait(&status)) != cpid && wpid != -1)
	;

    return (WIFEXITED(status) ? WEXITSTATUS(status) : -2);
}


/* SUSP -- Suspend the process.  Restore wrttmp entries first, since we can't
 * assume that stopped processes will be restarted in the reverse order in
 * which they were stopped.
 */

#ifdef SIGTSTP
RETSIGTYPE susp()
{
int mask;

	/* Disable further signals */
	mask= sigblock(sigmask(SIGTSTP));

	/* Restore modes and make wrttmp look like we exited */
	if (fixed_modes)
	{
		lseek(wstream,mypos,0);
		write(wstream,&mywrt,sizeof(struct wrttmp));
#ifdef TTYPERMS
		if (tmp_mesg != 's') resetperms();
#endif /*TTYPERMS*/
		fixed_modes=0;
	}

	signal(SIGTSTP,SIG_DFL);
	sigsetmask(0);
	kill(getpid(),SIGTSTP);

	/* STOP HERE */

	sigsetmask(mask);
	signal(SIGTSTP,(void (*)())susp);

	/* Reinstate wrttmp entry */
	lseek(wstream,mypos,0);
	write(wstream,&newwrt,sizeof(struct wrttmp));

#ifdef TTYPERMS
	if (tmp_mesg != 's')
	{
	    if (saveperms()) exit(1);
	    setperms(tmp_mesg);
	}
#endif /*TTYPERMS*/
}
#endif /*SIGTSTP*/
